# iDAM-UdeA

## Dataton Bancolombia 2022

By Antonio Tamayo

In [1]:
!pip install transformers
!pip install sentencepiece

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 5.3 MB 25.6 MB/s 
     |████████████████████████████████| 7.6 MB 30.9 MB/s 
     |████████████████████████████████| 163 kB 56.0 MB/s 
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 1.3 MB 22.4 MB/s 


## Cargar dataset (3 archivos csv)

In [2]:
#TODO

## Topic modeling con Zero-Shot

In [3]:
from transformers import pipeline

## Ejemplo con el primer artículo (título y artículo concatenados en un único string)

In [4]:
classifier = pipeline("zero-shot-classification", model="MoritzLaurer/mDeBERTa-v3-base-mnli-xnli")
sequence_to_classify = "Precio dolar hoy: la cotizacion de la divisa aumento y supero los $4.300. Este martes, 2 de agosto, el dolar alcanzo un precio maximo de $4.328 y un minimo de $4.245. Ademas, se negociaron US$889,35 millones en 1.650 transacciones. Un cambio en la divisa que representa una subia de $68,55 frente a la Tasa Representativa del Mercado (TRM hoy), que es de $4.314,54. TRM hoy: siga la cotizacion del dolar, en vivo, hoy viernes 22 de julio Le puede interesar: Cinco recomendaciones para hacerle el quite al precio del dolar y a la inflacion Escuche el podcast El Mundo Hoy"
candidate_labels = ["Macroeconomía", "sostenibilidad", "innovación", "regulaciones", "alianzas", "reputación"]
output = classifier(sequence_to_classify, candidate_labels, multi_label=False)
#print(output)
#Results
for l, s in zip(output['labels'], output['scores']):
  print(l,'\t',s)

Downloading:   0%|          | 0.00/1.07k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/558M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/463 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/4.31M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/18.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/156 [00:00<?, ?B/s]

/usr/local/lib/python3.7/dist-packages/transformers/convert_slow_tokenizer.py:447: UserWarning: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.
  "The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option"
/usr/local/lib/python3.7/dist-packages/transformers/models/deberta_v2/modeling_deberta_v2.py:746: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scale, dtype=query_layer.dtype
/usr/local/lib/python3.7/dist-packages/transformers/models/deberta_v2/modeling_deberta_

Macroeconomía 	 0.3853965103626251
innovación 	 0.15772944688796997
regulaciones 	 0.14235123991966248
alianzas 	 0.13817332684993744
reputación 	 0.10943428426980972
sostenibilidad 	 0.06691525131464005


## La noticia habla de un cliente en específico?

In [5]:
classifier = pipeline("zero-shot-classification", model="MoritzLaurer/mDeBERTa-v3-base-xnli-multilingual-nli-2mil7")
sequence_to_classify = 'El centro contribuye de manera relevante a nuestro proposito de poner al alcance de todos las oportunidades de esta nueva era, al promover la educacion financiera no solo como un tema relevante para el bienestar de las personas, sino tambien como una herramienta esencial para promover un crecimiento sostenible e inclusivo y asi crear oportunidades", ha manifestado Javier Rodriguez Soler en su primera intervencion ante los miembros del Consejo Asesor. Rodriguez Soler ha destacado el compromiso de los expertos que integran el centro. "Es un placer ver que, despues de estos anos, seguis apoyando al centro y nuestra ambicion compartida de promover la educacion financiera", ha senalado. En marzo de 2022, BBVA presento su Plan Global de Educacion Financiera con los objetivos de formar a un millon de personas en conocimientos y habilidades financieras; llegar a 50 millones de personas con contenidos de educacion financiera relevantes y gratuitos; desarrollar una formacion especifica para los clientes; y fomentar la cooperacion. Con este plan se refuerza la apuesta de BBVA por la educacion financiera como un elemento clave para mejorar la salud financiera de las personas y contribuir a la transicion hacia una economia mas sostenible, prioridades estrategicas de la entidad. Javier Rodriguez Soler destaco que el centro es una herramienta fundamental para cumplir estos objetivos y anuncio que en los proximos anos pondra el foco en "destacar el papel clave de la educacion financiera para promover un crecimiento sostenible e inclusivo''. Ademas anadio que seguira generando contenido en su plataforma ademas de difundir otros conocimientos practicos de educacion y salud financiera a traves de sus redes sociales y del programa Edufin Podcast y continuara siendo un referente internacional para el debate e intercambio de ideas. Durante la reunion se han repasado los principales indicadores de desempeno del centro que sirve de repositorio para casi 700 documentos de investigacion y buenas practicas y cuenta con mas de 15.000 usuarios unicos cada ano. Hasta la fecha el centro ha colaborado con mas de 100 entidades y en 22 proyectos de investigacion, y ha organizado eventos a los que han acudido mas de 4.000 personas en formato presencial o virtual.'
#IMPORTANTE: el modelo solo permite 10 labels al tiempo y se tienen 1058 clientes, por lo que se debe iterar este proceso tomando listas de 10 clientes a la vez.
candidate_labels = ["SUPERMERCADO LA GRAN COLOMBIA SA", "CAMAGUEY SA", "CARNES CASABLANCA S A", "SEATECH INTERNATIONAL INC", "GRUPO ALIMENTARIO DEL ATLANTICO SA", "CONGELADOS AGRICOLAS SA CONGELAGRO SA", "OLEOFLORES SAS", "PALMAS DEL CESAR SA", "BBVA", "Banco W"]
output = classifier(sequence_to_classify, candidate_labels, multi_label=False)
#print(output)
#Results
for l, s in zip(output['labels'], output['scores']):
  print(l,'\t',s)

Downloading:   0%|          | 0.00/1.09k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/558M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/467 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/4.31M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/16.3M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/23.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/173 [00:00<?, ?B/s]

BBVA 	 0.9272829294204712
Banco W 	 0.018894104287028313
PALMAS DEL CESAR SA 	 0.013588891364634037
CONGELADOS AGRICOLAS SA CONGELAGRO SA 	 0.01072651892900467
CAMAGUEY SA 	 0.00863668229430914
GRUPO ALIMENTARIO DEL ATLANTICO SA 	 0.007311627268791199
SUPERMERCADO LA GRAN COLOMBIA SA 	 0.006172223947942257
SEATECH INTERNATIONAL INC 	 0.003906896337866783
OLEOFLORES SAS 	 0.0019992555025964975
CARNES CASABLANCA S A 	 0.0014808273408561945


## Tareas

Preliminares:

Coordinar con los profesores Michal y Alexis la lectura y marcación de una porción del dataset para hacer una validación interna de los modelos.

Alternativas de solución

Alternativa 1.

1. Identificar el cliente:

  1.1 Lowercase a la lista de clientes
  1.2 Split a cada nombre de cliente por espacio
  1.3 Tokenize y lowercase a cada noticia
  1.4 Buscar si algún cliente se menciona en las noticias. Si el nombre del cliente es multipalabra, buscar la aparición de alguna de las palabras del nombre. Si aparece más de una coincidencia del nombre (palabra contenida en él) dentro de una misma noticia, tomar aquel que más veces aparezca. Si no hay moda, tomar el primero que se encuentre.

2. Realizar la categorización para todas las noticias del dataset y entregar la salida en el formato csv requerido en la competencia.

3. Para el ranking, tener en cuenta el peso que le da el banco a las categorías.

4. Combinar el peso de las categorías con el peso del sentimiento de la noticia. Para extraer el sentimiento de la noticia usar PySentimiento de Hugging Face y extraer la probabilidad del sentimiento.

ALternativa 2.

1. Usar Latent Dirichlet Allocation. Consultar la librería Gensim para esto.

Alternativa 3.

1. Usar K-means o clustering jerárquico aglomerativo con sklearn y Bag-of-Words, TF-IDF o word embeddings (preentranados o entrenarlos con word2vec).

Alternativa 4. 

Usar Eurowordnet para recuperar synsets de las categorías solicitadas. Categorizar por mayor frecuencia de aparición de sinónimos de las categorías.

Alternativa complementaria

Aplicar reglas con expresiones regulares sobre patrones útiles que se puedan identificar con la lectura de una porción del dataset.